File: frontend/src/lib/production/style.ts
import type { RendererConfig } from '../presentation/types.ts';
import type {
  BrandAsset,
  FontFamilyId,
  ParticipantBranding,
  ProductionStyle,
  Side
} from '../types.ts';

/**
 * Font stacks the renderer is allowed to use.
 *
 * These are local families only. Fetching a webfont at render time would make an export
 * depend on the network, and shipping font files would mean redistributing typefaces this
 * project has no licence to. A user who needs an exact typeface uploads it as a font
 * asset; see `docs/ASSETS.md` for the licensing responsibility that carries.
 */
export const FONT_STACKS: Record<Exclude<FontFamilyId, 'custom'>, string> = {
  system: 'system-ui, -apple-system, "Segoe UI", Roboto, sans-serif',
  geometric: '"Avenir Next", "Century Gothic", Futura, "Trebuchet MS", system-ui, sans-serif',
  grotesk: '"Helvetica Neue", Helvetica, Arial, system-ui, sans-serif',
  serif: '"Iowan Old Style", Georgia, "Times New Roman", serif',
  mono: 'ui-monospace, SFMono-Regular, Menlo, Consolas, monospace',
  // No pixel typeface is bundled, so Retro leans on a blocky monospace stack. Uploading a
  // real pixel font as a display asset replaces this without any code change.
  pixel: 'Monaco, "Andale Mono", "Courier New", ui-monospace, monospace'
};

export const P1_FALLBACK = '#6fffa8';
export const P2_FALLBACK = '#e36fff';

/** Neutral generated marks. KoalaBattle bundles no third-party logo files. */
export const MARK_LABELS: Record<string, string> = {
  gpt: 'GPT',
  gemini: 'GEMINI',
  claude: 'CLAUDE',
  deepseek: 'DEEPSEEK',
  local: 'LOCAL',
  manual: 'MANUAL',
  random: 'RANDOM',
  koala: 'KOALA'
};

/**
 * The Koala Broadcast defaults, mirroring `koalabattle.production.style_presets`.
 *
 * The API always sends a style, so this is a safety net for a stale cached response or a
 * hand-constructed timeline — not a second source of truth to edit presets in.
 */
export function defaultProductionStyle(): ProductionStyle {
  return {
    schema_version: '1.0',
    id: 'koala-broadcast',
    display_name: 'Koala Broadcast',
    version: '1.0',
    builtin: true,
    title: null,
    show_format: true,
    show_generation: true,
    show_koala_branding: true,
    stage: {
      background: {
        kind: 'arena', color: '#0b1f24', secondary_color: '#06090f', asset_id: null,
        fit: 'cover', position: 'center', brightness: 1, contrast: 1, blur: 0,
        overlay_opacity: 0, vignette: 0.35
      },
      arena: 'grid', floor_visible: true, ground_shadow: true, stage_lighting: 0.6,
      ambient_intensity: 0.6, background_motion: true, accent: '#7dffae'
    },
    hud: {
      preset: 'broadcast', hp_shape: 'slash', hp_thickness: 31, damage_ghost: true,
      show_hp_percent: true, show_hp_exact: true, show_level: false, show_types: true,
      show_status: true, show_player_name: true, show_provider: true, show_logo: true,
      show_player_slot: false, team_indicators: 'full', show_turn: true, show_weather: true
    },
    typography: {
      display: 'system', body: 'system', mono: 'mono', display_asset_id: null,
      body_asset_id: null, scale: 1, display_weight: 950, letter_spacing: 0,
      uppercase: true, outline: false, shadow: true
    },
    move: { layout: 'banner', show_type: true, show_archetype: true, duration_scale: 1 },
    damage: {
      show_damage: true, show_healing: true, show_effectiveness: true, show_critical: true,
      show_miss: true, show_immune: true, intensity: 'standard'
    },
    commentary: {
      layout: 'fighter-card', show_agent_name: true, show_logo: true, show_label: true,
      animation: 'fade'
    },
    caption: {
      preset: 'broadcast', show_speaker: false, background_opacity: 0.88, outline: false,
      size_scale: 1, position: 'bottom'
    },
    effect: {
      intensity: 'standard', camera: 'subtle', idle_motion: 'full', pacing: 'standard',
      impact_flash: true, trails: true
    },
    intro: {
      enabled: true, length: 'standard', show_player_logos: true, show_player_names: true,
      show_format: true, show_generation: true, show_game_number: false,
      show_series_score: false, show_tournament_round: false
    },
    result: {
      enabled: true, show_winner: true, show_logos: true, show_final_score: false,
      show_format: true, show_series: false, duration_ms: 3600
    },
    watermark: {
      enabled: false, asset_id: null, text: null, position: 'bottom-right',
      opacity: 0.55, size: 1
    },
    players: {},
    series: {
      tournament_name: null, round_name: null, game_number: null, best_of: null,
      score_p1: null, score_p2: null
    }
  };
}

export function assetUrl(apiBase: string, assetId: string | null | undefined): string | null {
  return assetId ? `${apiBase}/api/branding/assets/${assetId}/media` : null;
}

export function fontFamily(
  style: ProductionStyle,
  role: 'display' | 'body' | 'mono'
): string {
  const typography = style.typography;
  const custom = role === 'display' ? typography.display_asset_id : role === 'body' ? typography.body_asset_id : null;
  const stack = FONT_STACKS[(typography[role] === 'custom' ? 'system' : typography[role]) as Exclude<FontFamilyId, 'custom'>];
  // A missing custom font must fall back visibly-but-sanely rather than silently
  // rendering one face in the preview and another in the export.
  return custom ? `"${customFontName(custom)}", ${stack}` : stack;
}

export function customFontName(assetId: string): string {
  return `kb-font-${assetId}`;
}

const registered = new Map<string, Promise<boolean>>();

/**
 * Register every custom font this style needs and wait for it.
 *
 * Called before the first frame is composited so a font can never load half way through a
 * render, which would make an export non-deterministic.
 */
export async function ensureStyleFonts(style: ProductionStyle, apiBase: string): Promise<string[]> {
  const wanted = [style.typography.display_asset_id, style.typography.body_asset_id].filter(
    (value): value is string => Boolean(value)
  );
  const results = await Promise.all(wanted.map((assetId) => loadFont(assetId, apiBase)));
  return wanted.filter((_, index) => !results[index]);
}

async function loadFont(assetId: string, apiBase: string): Promise<boolean> {
  const key = `${apiBase}|${assetId}`;
  let pending = registered.get(key);
  if (!pending) {
    pending = (async () => {
      if (typeof FontFace !== 'function' || typeof document === 'undefined') return false;
      try {
        const face = new FontFace(customFontName(assetId), `url(${assetUrl(apiBase, assetId)})`);
        await face.load();
        document.fonts.add(face);
        return true;
      } catch {
        return false;
      }
    })();
    registered.set(key, pending);
  }
  return pending;
}

export function brandingFor(style: ProductionStyle, side: Side): ParticipantBranding {
  return (
    style.players[side] || {
      display_name: null,
      short_name: null,
      logo_asset_id: null,
      logo_mark: null,
      accent: null,
      secondary_accent: null
    }
  );
}

export function accentFor(style: ProductionStyle, side: Side): string {
  return brandingFor(style, side).accent || (side === 'p1' ? P1_FALLBACK : P2_FALLBACK);
}

/** `Gen 9 · Random Battle`, never the raw `gen9randombattle` id. */
export function formatDisplayName(formatId: string, showGeneration = true): string {
  const normalized = formatId.toLowerCase().replaceAll('_', '').replaceAll('-', '');
  const parsed = normalized.match(/^gen(\d+)(.+)$/);
  if (!parsed) return titleCase(formatId.replaceAll('_', ' ').replaceAll('-', ' '));
  const tiers: Record<string, string> = {
    randombattle: 'Random Battle',
    randomdoublesbattle: 'Random Doubles',
    ou: 'OU',
    uu: 'UU',
    ru: 'RU',
    nu: 'NU',
    pu: 'PU',
    lc: 'LC',
    ubers: 'Ubers',
    monotype: 'Monotype',
    doublesou: 'Doubles OU',
    vgc2024: 'VGC 2024'
  };
  const tier = tiers[parsed[2]] || titleCase(parsed[2]);
  return showGeneration ? `Gen ${parsed[1]} · ${tier}` : tier;
}

export function generationOf(formatId: string): number {
  const parsed = formatId.toLowerCase().match(/^gen(\d)/);
  return parsed ? Number(parsed[1]) : 9;
}

function titleCase(value: string): string {
  return value.replace(/\b[a-z]/g, (character) => character.toUpperCase());
}

/** How much a style's motion settings scale visual intensity. */
export function intensityScale(value: string): number {
  return { off: 0, minimal: 0.4, standard: 1, dramatic: 1.45 }[value] ?? 1;
}

export function idleScale(value: string): number {
  return { off: 0, subtle: 0.45, full: 1 }[value] ?? 1;
}

export function cameraScale(value: string): number {
  return { static: 0, subtle: 1, dynamic: 1.9 }[value] ?? 1;
}

export function assetById(assets: BrandAsset[], id: string | null): BrandAsset | null {
  return id ? assets.find((asset) => asset.id === id) || null : null;
}

/**
 * Project a ProductionStyle onto the live DOM renderer's settings.
 *
 * There is one style model, not one per surface. The offline compositor consumes the
 * whole style; the live `/watch` and `/overlay` surfaces are a DOM renderer with a
 * smaller vocabulary, so the style is *mapped* here rather than reimplemented. Anything
 * the DOM renderer cannot express is simply not applied — it is never approximated with
 * a second set of settings.
 */
export function styleToRendererConfig(
  style: ProductionStyle,
  base: RendererConfig
): RendererConfig {
  const effects =
    style.effect.intensity === 'off'
      ? 'off'
      : style.effect.intensity === 'minimal'
        ? 'low'
        : style.effect.intensity === 'dramatic'
          ? 'high'
          : 'standard';
  return {
    ...base,
    effects,
    reducedMotion: style.effect.idle_motion === 'off' || style.effect.camera === 'static',
    showDamageNumbers: style.damage.show_damage,
    showTurn: style.hud.show_turn,
    commentaryMode: style.commentary.layout === 'off' ? 'hidden' : base.commentaryMode
  };
}

File: backend/koalabattle/production/style.py
"""Declarative, versioned presentation settings for a production.

A :class:`ProductionStyle` describes *how* a replay should look and sound. It never
describes *what* happened: battle events, decisions, teams and the winner are owned by the
match archive and are untouched by anything in this module. One match may therefore carry
any number of productions, each with its own style, and rendering one can never change
another.

Every field here has to survive the trip to the native offline compositor, so the model
stays declarative — bounded enumerations and numbers, never CSS, markup or file paths.
Assets are referenced by generated :class:`~koalabattle.branding.models.BrandAsset` ids.
"""

from __future__ import annotations

from enum import StrEnum
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field

STYLE_SCHEMA_VERSION = "1.0"

#: A CSS-safe colour. Only `#rgb` / `#rrggbb` is accepted so a style can never smuggle
#: arbitrary CSS (`url(...)`, `expression(...)`) into a rendering surface.
Color = Annotated[str, Field(pattern=r"^#(?:[0-9a-fA-F]{3}|[0-9a-fA-F]{6})$")]
AssetId = Annotated[str, Field(pattern=r"^[a-z0-9]{32}$")]
PresetId = Annotated[str, Field(pattern=r"^[a-z0-9][a-z0-9-]{1,58}[a-z0-9]$")]


class FrozenStyleModel(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)


class Intensity(StrEnum):
    OFF = "off"
    MINIMAL = "minimal"
    STANDARD = "standard"
    DRAMATIC = "dramatic"


class FontFamily(StrEnum):
    """Curated font choices.

    These are local stacks rather than downloaded webfonts: the offline renderer must not
    depend on a network fetch, and KoalaBattle must not redistribute fonts it has no
    licence to ship. Users who want a specific typeface add it themselves as a font asset.
    """

    SYSTEM = "system"
    GEOMETRIC = "geometric"
    GROTESK = "grotesk"
    SERIF = "serif"
    MONO = "mono"
    PIXEL = "pixel"
    CUSTOM = "custom"


class BackgroundStyle(FrozenStyleModel):
    kind: Literal["arena", "solid", "gradient", "image"] = "arena"
    color: Color = "#0b1f24"
    secondary_color: Color = "#06090f"
    asset_id: AssetId | None = None
    fit: Literal["cover", "contain"] = "cover"
    position: Literal["center", "top", "bottom", "left", "right"] = "center"
    brightness: float = Field(default=1.0, ge=0.2, le=1.6)
    contrast: float = Field(default=1.0, ge=0.5, le=1.8)
    blur: int = Field(default=0, ge=0, le=40)
    overlay_opacity: float = Field(default=0.0, ge=0.0, le=0.9)
    vignette: float = Field(default=0.35, ge=0.0, le=1.0)


class StageStyle(FrozenStyleModel):
    background: BackgroundStyle = Field(default_factory=BackgroundStyle)
    arena: Literal["none", "stadium", "platform", "minimal-floor", "grid"] = "grid"
    floor_visible: bool = True
    ground_shadow: bool = True
    stage_lighting: float = Field(default=0.6, ge=0.0, le=1.0)
    ambient_intensity: float = Field(default=0.6, ge=0.0, le=1.0)
    background_motion: bool = True
    accent: Color = "#7dffae"


class HudStyle(FrozenStyleModel):
    preset: Literal["broadcast", "fighting", "minimal", "esports", "retro"] = "broadcast"
    hp_shape: Literal["slash", "rounded", "square", "pill"] = "slash"
    hp_thickness: int = Field(default=31, ge=8, le=54)
    damage_ghost: bool = True
    show_hp_percent: bool = True
    show_hp_exact: bool = True
    show_level: bool = False
    show_types: bool = True
    show_status: bool = True
    show_player_name: bool = True
    show_provider: bool = True
    show_logo: bool = True
    show_player_slot: bool = False
    #: ``revealed`` only shows opponent Pokemon the spectator has already seen. Nothing in
    #: this enum can widen visibility beyond the public presentation archive.
    team_indicators: Literal["full", "revealed", "fainted-only", "hidden"] = "full"
    show_turn: bool = True
    show_weather: bool = True


class TypographyStyle(FrozenStyleModel):
    display: FontFamily = FontFamily.SYSTEM
    body: FontFamily = FontFamily.SYSTEM
    mono: FontFamily = FontFamily.MONO
    display_asset_id: AssetId | None = None
    body_asset_id: AssetId | None = None
    scale: float = Field(default=1.0, ge=0.8, le=1.3)
    display_weight: int = Field(default=950, ge=400, le=950)
    letter_spacing: float = Field(default=0.0, ge=-2.0, le=6.0)
    uppercase: bool = True
    outline: bool = False
    shadow: bool = True


class MoveCalloutStyle(FrozenStyleModel):
    layout: Literal["banner", "impact", "minimal", "lower-third", "centered", "off"] = "banner"
    show_type: bool = True
    show_archetype: bool = True
    duration_scale: float = Field(default=1.0, ge=0.5, le=2.0)


class DamageStyle(FrozenStyleModel):
    show_damage: bool = True
    show_healing: bool = True
    show_effectiveness: bool = True
    show_critical: bool = True
    show_miss: bool = True
    show_immune: bool = True
    intensity: Intensity = Intensity.STANDARD


class CommentaryStyle(FrozenStyleModel):
    layout: Literal["fighter-card", "side-panel", "lower-third", "bubble", "caption", "off"] = (
        "fighter-card"
    )
    show_agent_name: bool = True
    show_logo: bool = True
    show_label: bool = True
    animation: Literal["fade", "slide", "punch", "minimal", "none"] = "fade"


class CaptionStyle(FrozenStyleModel):
    preset: Literal["broadcast", "minimal", "high-contrast", "vertical", "off"] = "broadcast"
    show_speaker: bool = False
    background_opacity: float = Field(default=0.88, ge=0.0, le=1.0)
    outline: bool = False
    size_scale: float = Field(default=1.0, ge=0.7, le=1.5)
    position: Literal["bottom", "center", "top"] = "bottom"


class EffectStyle(FrozenStyleModel):
    intensity: Intensity = Intensity.STANDARD
    camera: Literal["static", "subtle", "dynamic"] = "subtle"
    idle_motion: Literal["full", "subtle", "off"] = "full"
    pacing: Literal["cinematic", "standard", "fast"] = "standard"
    impact_flash: bool = True
    trails: bool = True


class IntroStyle(FrozenStyleModel):
    enabled: bool = True
    length: Literal["quick", "standard", "dramatic"] = "standard"
    show_player_logos: bool = True
    show_player_names: bool = True
    show_format: bool = True
    show_generation: bool = True
    show_game_number: bool = False
    show_series_score: bool = False
    show_tournament_round: bool = False

    @property
    def duration_ms(self) -> int:
        return {"quick": 1600, "standard": 3200, "dramatic": 5000}[self.length]


class ResultStyle(FrozenStyleModel):
    enabled: bool = True
    show_winner: bool = True
    show_logos: bool = True
    show_final_score: bool = False
    show_format: bool = True
    show_series: bool = False
    duration_ms: int = Field(default=3600, ge=800, le=12_000)


class WatermarkStyle(FrozenStyleModel):
    enabled: bool = False
    asset_id: AssetId | None = None
    text: str | None = Field(default=None, max_length=40)
    position: Literal["top-left", "top-right", "bottom-left", "bottom-right"] = "bottom-right"
    opacity: float = Field(default=0.55, ge=0.05, le=1.0)
    size: float = Field(default=1.0, ge=0.5, le=2.0)


class ParticipantBranding(FrozenStyleModel):
    """How one player is presented. Deliberately independent of which provider runs them.

    A Manual Web Chat player driven by a human pasting into ChatGPT may legitimately be
    presented as "ChatGPT"; a generic OpenAI-compatible endpoint may be presented as
    "My Local Model". Provider implementation and on-screen identity are separate concerns.
    """

    display_name: str | None = Field(default=None, min_length=1, max_length=40)
    short_name: str | None = Field(default=None, min_length=1, max_length=12)
    logo_asset_id: AssetId | None = None
    #: A bundled neutral provider mark (see ``koalabattle.branding.marks``).
    logo_mark: str | None = Field(default=None, max_length=24)
    accent: Color | None = None
    secondary_accent: Color | None = None


class SeriesDisplay(FrozenStyleModel):
    tournament_name: str | None = Field(default=None, max_length=60)
    round_name: str | None = Field(default=None, max_length=40)
    game_number: int | None = Field(default=None, ge=1, le=99)
    best_of: int | None = Field(default=None, ge=1, le=99)
    score_p1: int | None = Field(default=None, ge=0, le=99)
    score_p2: int | None = Field(default=None, ge=0, le=99)


class ProductionStyle(FrozenStyleModel):
    schema_version: str = STYLE_SCHEMA_VERSION
    id: PresetId = "koala-broadcast"
    display_name: str = Field(default="Koala Broadcast", min_length=1, max_length=60)
    version: str = Field(default="1.0", max_length=20)
    builtin: bool = True
    title: str | None = Field(default=None, min_length=1, max_length=90)
    show_format: bool = True
    show_generation: bool = True
    show_koala_branding: bool = True
    stage: StageStyle = Field(default_factory=StageStyle)
    hud: HudStyle = Field(default_factory=HudStyle)
    typography: TypographyStyle = Field(default_factory=TypographyStyle)
    move: MoveCalloutStyle = Field(default_factory=MoveCalloutStyle)
    damage: DamageStyle = Field(default_factory=DamageStyle)
    commentary: CommentaryStyle = Field(default_factory=CommentaryStyle)
    caption: CaptionStyle = Field(default_factory=CaptionStyle)
    effect: EffectStyle = Field(default_factory=EffectStyle)
    intro: IntroStyle = Field(default_factory=IntroStyle)
    result: ResultStyle = Field(default_factory=ResultStyle)
    watermark: WatermarkStyle = Field(default_factory=WatermarkStyle)
    players: dict[str, ParticipantBranding] = Field(default_factory=dict)
    series: SeriesDisplay = Field(default_factory=SeriesDisplay)

    def asset_ids(self) -> tuple[str, ...]:
        """Every brand asset this style needs, for preflight and manifest snapshots."""
        candidates = [
            self.stage.background.asset_id,
            self.typography.display_asset_id,
            self.typography.body_asset_id,
            self.watermark.asset_id,
            *(branding.logo_asset_id for branding in self.players.values()),
        ]
        seen: dict[str, None] = {}
        for candidate in candidates:
            if candidate:
                seen.setdefault(candidate, None)
        return tuple(seen)

    def branding_for(self, side: str) -> ParticipantBranding:
        return self.players.get(side, ParticipantBranding())


class StylePreset(FrozenStyleModel):
    """A saved, reusable style. Built-in presets are read-only; users duplicate them."""

    id: PresetId
    display_name: str = Field(min_length=1, max_length=60)
    description: str = Field(default="", max_length=200)
    builtin: bool = False
    style: ProductionStyle
    created_at: str | None = None
    updated_at: str | None = None


class SaveStylePreset(FrozenStyleModel):
    display_name: str = Field(min_length=1, max_length=60)
    description: str = Field(default="", max_length=200)
    style: ProductionStyle


class UpdateProductionStyle(FrozenStyleModel):
    style: ProductionStyle
